sklearn/inspection/_partial_dependence.py

```diff
@@ -349,21 +368,44 @@ def partial_dependence(estimator, X, features, response_method='auto',
                 "'decision_function'. Got {}.".format(response_method)
             )
 
-    n_features = X.shape[1]
-    features = np.asarray(features, dtype=np.int32, order='C').ravel()
-    if any(not (0 <= f < n_features) for f in features):
-        raise ValueError('all features must be in [0, %d]'
-                         % (n_features - 1))
+    if isinstance(features, slice):
```

Comment:
Isn't there a way to raise those informative error message for any unsupported "features" type instead of just for instances of the `slice` type?

Reply:
Hum, I had a look at `_determine_key_type` which is also used by `ColumnTransformer` and it already accepts features that are of the  `slice` type. I had not realized this. It's too late to drop support for slices in  `ColumnTransformer` although I think this is a YAGNI.

So maybe we should also accept feature slices for PDP to stay consistent. WDYT @NicolasHug?

Reply:
Alright. Maybe we could extend `_determine_key_type` to add an `accept_slice=False` option directly in it.

Reply:
Yeah I was thinking about that too

Reply:
OK let's go with this addition